### PR TITLE
implementing benchpress

### PR DESCRIPTION
**Pull request by polatengin**
Replacing custom tests with BenchPress tests

Successful CI links [AZDO Bicep CI](https://dev.azure.com/csedevops/Symphony/_build/results?buildId=15554&view=logs&j=5264e576-3c6f-51f6-f055-fab409685f20) , [GH Bicep CI](https://github.com/microsoft/symphony/actions/runs/6579068500)

### CODE REVIEW

File: .mega-linter.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ ENABLE_LINTERS:
   - ENV_DOTENV_LINTER
   - YAML_YAMLLINT
   - JSON_JSONLINT
-  - ACTION_ACTIONLINT
+# - ACTION_ACTIONLINT
```
Comment (hattan) [inline .mega-linter.yml:25]: Not a blocker: can you open an issue to add actionlint back in and reference it in this pr. Also consider a comment explaining why actionlint was disabled , with the intention of having a separate pr to resolve the bash linting issues that is part of the the run action.

File: scripts/orchestrators/tests.runner.sh [hunk 2]
```diff
@@ -118,7 +118,7 @@ bicep() {
   }
 
   # cd to the tests directory
-  pushd ../../IAC/Bicep/test
+  pushd ../../IAC/Bicep/tests
```
Comment (hattan) [inline scripts/orchestrators/tests.runner.sh:121]: Can we please revert the test folder name back to test (singular) so we have parity with Terraform?

File: scripts/orchestrators/scanners.sh [hunk 3]
```diff
@@ -42,6 +42,6 @@ run_armttk() {
     local bicep_file_path=$1
 
     az bicep build --file "${bicep_file_path}"
-    arm-ttk/Test-AzTemplate.sh "${bicep_file_path/.bicep/.json}"
+    arm-ttk/arm-ttk/Test-AzTemplate.sh "${bicep_file_path/.bicep/.json}"
```
Comment (hattan) [inline scripts/orchestrators/scanners.sh:45]: Is there a reason to have arm-ttk twice in the path?

File: IAC/Bicep/tests/spec/support/custom_matcher.sh [hunk 4]
```diff
@@ -6,8 +6,8 @@ shellspec_matcher_include_name() {
     shellspec_matcher__match() {
         SHELLSPEC_EXPECT="$1"
         SHELLSPEC_SUBJECT=$(echo "$SHELLSPEC_SUBJECT" | jq -r '.name')
-        [ "$SHELLSPEC_EXPECT" == "$SHELLSPEC_EXPECT" ] || return 1
-        expr "$SHELLSPEC_SUBJECT" : "$SHELLSPEC_EXPECT" >/dev/null || return 1
+        [ "$SHELLSPEC_EXPECT" = "$SHELLSPEC_EXPECT" ] || return 1
```
Comment (hattan) [inline IAC/Bicep/tests/spec/support/custom_matcher.sh:9]: Can we revert these changes please? They are part of the supporting matcher code for shellspec (which isn't used.) 

We should add a story to rmove shellspec or move this implementation to the docs folder under an alternative. Especially since the tests are no longer at part with the BenchPress tests.

**Review by hattan — CHANGES_REQUESTED** [review]
Thanks for the PR, @polatengin . Can you please also include links to CI runs for Bicep on Both GitHub actions and AzDo?

Thanks!

File: scripts/orchestrators/tests.runner.sh (hunk 2 above)
Comment (polatengin) [reply scripts/orchestrators/tests.runner.sh]: good point!

File: IAC/Bicep/tests/spec/support/custom_matcher.sh (hunk 4 above)
Comment (polatengin) [reply IAC/Bicep/tests/spec/support/custom_matcher.sh]: fair enough, reverted changes back 👍

**Comment by polatengin** [comment]
you can find an example execution here; https://dev.azure.com/enpolat/symphony_test_2/_build/results?buildId=573&view=results

File: .mega-linter.yml (hunk 1 above)
Comment (polatengin) [reply .mega-linter.yml]: yes 👍 quick note; I fixed an issue related to actionlinter here; https://github.com/microsoft/symphony/pull/234

File: scripts/orchestrators/scanners.sh (hunk 3 above)
Comment (polatengin) [reply scripts/orchestrators/scanners.sh:45]: that's the correct path, you can find the successful execution logs here; https://dev.azure.com/enpolat/symphony_test_2/_build/results?buildId=573&view=results

File: .github/workflows/template.bicep.test.yml [hunk 5]
```diff
@@ -72,10 +73,15 @@ jobs:
...
-          export ARM_CLIENT_SECRET=${{ steps.LoadSecrets.outputs.clientSecret }}
+          export ARM_TENANT_ID="${{ steps.LoadSecrets.outputs.tenantId }}"
+          export ARM_SUBSCRIPTION_ID="${{ steps.LoadSecrets.outputs.subscriptionId }}"
+          export ARM_CLIENT_ID="${{ steps.LoadSecrets.outputs.clientId }}"
+          export ARM_CLIENT_SECRET="${{ steps.LoadSecrets.outputs.clientSecret }}"
+
+          export AZ_TENANT_ID="${{ steps.LoadSecrets.outputs.tenantId }}"
+          export AZ_SUBSCRIPTION_ID="${{ steps.LoadSecrets.outputs.subscriptionId }}"
+          export AZ_APPLICATION_ID="${{ steps.LoadSecrets.outputs.clientId }}"
+          export AZ_ENCRYPTED_PASSWORD="$(pwsh -c "\"${{ steps.LoadSecrets.outputs.clientSecret }}\" | ConvertTo-SecureString -AsPlainText -Force | ConvertFrom-SecureString")"
```
Comment (PabloZaiden) [inline .github/workflows/template.bicep.test.yml:84]: Just to be extra safe, I'd also mask the encrypted password from the pipeline output

File: scripts/orchestrators/setup-benchpress.sh [hunk 6]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
```
Comment (PabloZaiden) [inline scripts/orchestrators/setup-benchpress.sh:1]: Instead of constantly jumping to `pwsh` shells, wouldn't it be better to have the whole benchpress setup logic in a powershell file and just invoke one `pwsh` command?

File: .github/workflows/template.bicep.test.yml (hunk 5 above)
Comment (polatengin) [reply .github/workflows/template.bicep.test.yml:84]: We are only setting environment variables, and they do not appear in the terminal logs. Therefore, there's no need to mask them as they are not logged in to the session.

File: scripts/orchestrators/setup-benchpress.sh (hunk 6 above)
Comment (polatengin) [reply scripts/orchestrators/setup-benchpress.sh:1]: `setup-xxx` scripts are all in bash format, I don't want to make an exception for benchpress and have it in powershell format.

in the `setup-benchpress.sh`, we do checks to find,

- if benchpress itself is installed or not
- if all dependencies are installed or not

if they're not installed, they get installed by the script, if they already installed, we just log it and continue

if we merge all powershell scripts into one, we can't check individual dependencies are installed or not

Comment (PabloZaiden) [reply scripts/orchestrators/setup-benchpress.sh:1]: Yeah, I get that. What I was thinking was that `setup-benchpress.sh` could be just a bootstrapper for a `setup-benchpress.ps1` and you can run everything from that single powershell script, instead of keep calling `pwsh`. Because right now you're invoking `pwsh` to check stuff, but the logic itself is in `bash`. Don't love the mix, and that's why I thought of having a single script that does the whole benchpress setup would be a better idea.

Comment (polatengin) [reply scripts/orchestrators/setup-benchpress.sh:1]: make sense, done!

**Review by PabloZaiden — APPROVED** [review]
Looks good.

File: .github/workflows/template.bicep.test.yml [hunk 7]
```diff
@@ -43,6 +43,7 @@ jobs:
         with:
           creds: "${{ secrets.azureEnviromentCredentials }}"
           environment: "AzureCloud"
+          enable-AzPSSession: true
```
Comment (PabloZaiden) [inline .github/workflows/template.bicep.test.yml:46]: *this* was the trick we were missing. Good catch!

Comment (polatengin) [reply .github/workflows/template.bicep.test.yml:46]: 👍

**Review by HadwaAbdelhalem — APPROVED** [review]
@polatengin I do not see the unit/modules tests integrated in the pipelines. this is a gap that we currently have and it would be good to fix it in the scope of this PR. can you please run those tests as part of the validate stage based on a flag similar to what we have in the terraform.

**Comment by polatengin** [comment]
> @polatengin I do not see the unit/modules tests integrated in the pipelines. this is a gap that we currently have and it would be good to fix it in the scope of this PR. can you please run those tests as part of the validate stage based on a flag similar to what we have in the terraform.

thanks for the comment @HadwaAbdelhalem. I created an issue here; https://github.com/microsoft/symphony/issues/236, I'll create a follow-up PR to address it 👍

**Review by hattan — APPROVED** [review]
LGTM
